flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/DataTypeExtractor.java

```diff
@@ -0,0 +1,574 @@
+		final List<Type> structuredTypeHierarchy = collectTypeHierarchy(Object.class, type);
+		for (Field field : fields) {
+			try {
+				final Type fieldType = field.getGenericType();
+				final List<Type> fieldTypeHierarchy = new ArrayList<>();
+				// hierarchy until structured type
+				fieldTypeHierarchy.addAll(typeHierarchy);
+				// hierarchy of structured type
+				fieldTypeHierarchy.addAll(structuredTypeHierarchy);
+				// field type
+				fieldTypeHierarchy.add(fieldType);
+				final DataTypeTemplate fieldTemplate = mergeFieldTemplate(lookup, field, template);
+				final DataType fieldDataType = extractDataTypeOrRaw(fieldTemplate, fieldTypeHierarchy, fieldType);
+				fieldDataTypes.put(field.getName(), fieldDataType);
+			} catch (Throwable t) {
+				throw extractionError(
+					t,
+					"Error in field '%s' of class '%s'.",
+					field.getName(),
+					field.getDeclaringClass().getName());
+			}
+		}
+		return fieldDataTypes;
+	}
+
+	private List<StructuredAttribute> createStructuredTypeAttributes(
+			ExtractionUtils.AssigningConstructor constructor,
+			Map<String, DataType> fieldDataTypes) {
+		// field order is defined by assigning constructor
+		if (constructor != null) {
+			return constructor.parameterNames
+				.stream()
+				.map(name -> {
+					final LogicalType logicalType = fieldDataTypes.get(name).getLogicalType();
+					return new StructuredAttribute(name, logicalType);
+				})
+				.collect(Collectors.toList());
+		}
+		// field order is sorted
+		else {
+			return fieldDataTypes.keySet()
+				.stream()
+				.sorted()
+				.map(name -> {
+					final LogicalType logicalType = fieldDataTypes.get(name).getLogicalType();
+					return new StructuredAttribute(name, logicalType);
+				})
+				.collect(Collectors.toList());
+		}
+	}
+
+	/**
+	 * Merges the template of a structured type with a possibly more specific field annotation.
+	 */
+	private DataTypeTemplate mergeFieldTemplate(DataTypeLookup lookup, Field field, DataTypeTemplate structuredTemplate) {
+		final DataTypeHint hint = field.getAnnotation(DataTypeHint.class);
+		if (hint == null) {
+			return structuredTemplate.copyWithoutDataType();
+		}
+		return structuredTemplate.mergeWithInnerAnnotation(lookup, hint);
+	}
+
+	/**
+	 * Use closest class for data type if possible. Even though a hint might have provided some data
+	 * type, in many cases, the conversion class can be enriched with the extraction type itself.
+	 */
+	@SuppressWarnings("unchecked")
+	private DataType closestBridging(DataType dataType, @Nullable Class<?> clazz) {
+		if (clazz == null) {
+			return dataType;
+		}
+		final LogicalType logicalType = dataType.getLogicalType();
+		final boolean supportsConversion = logicalType.supportsInputConversion(clazz) ||
+			logicalType.supportsOutputConversion(clazz);
+		if (supportsConversion && logicalType instanceof RawType) {
+			return DataTypes.RAW(clazz, ((RawType) logicalType).getTypeSerializer());
```

Comment:
This is strange. Shouldn't the originating class stay intact?

Reply:
Not if we can use a more specific class. E.g. `eval(@DataTypeHint("RAW") Integer o)` the annotation itself says `Object.class` because no conversion class was specified but the context knows it better.

Reply:
Will it work if it is the other way round? `eval(@DataTypeHing("Raw", bridgedTo = Integer.class) Object o)`

Reply:
I changed the logic.